client/src/main.rs

```diff
@@ -1,32 +1,229 @@
+    /// followed by `"\n\r"`), then restore the prompt line. stdout is not flushed.
+    fn print_upwards<F>(&mut self, f: F) -> Result<()>
+    where
+        F: Fn() -> Result<()>,
+    {
+        self.stdout
+            .queue(terminal::Clear(ClearType::CurrentLine))?
+            .queue(cursor::MoveToColumn(0))?;
+        f()?;
+        print!("{}{}", PROMPT, self.input);
+        Ok(())
+    }
+
+    fn send_fifo_msg(&mut self, msg: &ClientToDaemonMsg) -> Result<()> {
+        rmp_serde::encode::write(&mut self.ctodbuf, &msg)
+            .with_context(|| format!("couldn't write message {msg:?} to fifo"))
+    }
+
+    fn handle_fifo_msg(&mut self) -> Result<()> {
+        let mut data = Vec::new();
+        self.dtocbuf
+            .read_to_end(&mut data)
+            .context("couldn't read from fifo")?;
+        match rmp_serde::from_slice::<DaemonToClientMsg>(&data) {
+            Ok(msg) => print!("got fifo msg: {msg:?}\n\r"),
+            Err(e) => self.print_upwards(|| {
+                bunt::print!("{$red}invalid message from daemon: {}{/$}\n\r", e);
+                Ok(())
+            })?,
+        }
+
+        Ok(())
+    }
+
+    fn handle_term_evt(&mut self) -> Result<ControlFlow<()>> {
+        let event = event::read().context("couldn't read next terminal event")?;
+        match event {
+            Event::Key(kev) => self.handle_kev(kev),
+            _ => Ok(ControlFlow::Continue(())),
+        }
+    }
+
+    fn handle_kev(&mut self, kev: KeyEvent) -> Result<ControlFlow<()>> {
+        match kev.code {
+            KeyCode::Char('c') if kev.modifiers == KeyModifiers::CONTROL => {
+                return Ok(ControlFlow::Break(()));
+            }
+            KeyCode::Backspace => {
+                if !self.input.is_empty() {
+                    self.input.pop();
+                    print!("\x08 \x08");
+                }
+            }
+            KeyCode::Enter => {
+                self.send_fifo_msg(&ClientToDaemonMsg::Send(self.input.clone()))?;
+                self.input.clear();
+                self.stdout
+                    .queue(terminal::Clear(ClearType::CurrentLine))?
+                    .queue(cursor::MoveToColumn(0))?;
+                print!("{}", PROMPT);
+            }
+            KeyCode::Char(c) => {
+                self.input.push(c);
+                print!("{c}");
+            }
+            _ => (),
+        }
+
+        Ok(ControlFlow::Continue(()))
     }
+}
 
-    loop {
-        let bytes = fs::read(&fifo).context("couldn't read fifo")?;
-        match rmp_serde::from_slice::<DaemonToClientMsg>(&bytes) {
-            Ok(msg) => println!("recieved message: {msg:?}"),
-            Err(e) => println!("recieved invalid message: {e}"),
+fn chat() -> Result<()> {
+    let mut state = State::new()?;
+    state.start()?;
+
+    Ok(())
+}
+
+#[allow(unused_must_use)]
+/// Try our best to clean up the terminal state; if too many errors happen, just print some newlines and call it good.
```

Comment:
Average frontend developer

Reply:
i copied that code almost exactly from `guac`